File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.java
package org.wordpress.android.ui.prefs.notifications;

import android.app.FragmentManager;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.preference.PreferenceManager;

import com.google.android.material.elevation.ElevationOverlayProvider;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.wordpress.android.R;
import org.wordpress.android.analytics.AnalyticsTracker;
import org.wordpress.android.ui.LocaleAwareActivity;
import org.wordpress.android.ui.notifications.NotificationEvents;
import org.wordpress.android.ui.prefs.notifications.PrefMasterSwitchToolbarView.MasterSwitchToolbarListener;
import org.wordpress.android.util.ContextExtensionsKt;

// Simple wrapper activity for NotificationsSettingsFragment
public class NotificationsSettingsActivity extends LocaleAwareActivity
        implements MasterSwitchToolbarListener {
    private TextView mMessageTextView;
    private View mMessageContainer;

    protected SharedPreferences mSharedPreferences;
    protected View mFragmentContainer;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.notifications_settings_activity);
        mFragmentContainer = findViewById(R.id.fragment_container);

        // Get shared preferences for master switch.
        mSharedPreferences = PreferenceManager.getDefaultSharedPreferences(NotificationsSettingsActivity.this);

        // Set up primary toolbar
        setUpToolbar();

        // Set up master switch
        setUpMasterSwitch();

        FragmentManager fragmentManager = getFragmentManager();
        if (savedInstanceState == null) {
            fragmentManager.beginTransaction()
                           .add(R.id.fragment_container, new NotificationsSettingsFragment())
                           .commit();
        }

        mMessageContainer = findViewById(R.id.notifications_settings_message_container);
        mMessageTextView = findViewById(R.id.notifications_settings_message);
    }

    @Override
    protected void onStop() {
        EventBus.getDefault().unregister(this);
        super.onStop();
    }

    @Override
    protected void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("unused")
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEventMainThread(NotificationEvents.NotificationsSettingsStatusChanged event) {
        if (TextUtils.isEmpty(event.getMessage())) {
            mMessageContainer.setVisibility(View.GONE);
        } else {
            mMessageContainer.setVisibility(View.VISIBLE);
            mMessageTextView.setText(event.getMessage());
        }
    }

    /**
     * Set up primary toolbar for navigation and search
     */
    private void setUpToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar_with_search);

        if (toolbar != null) {
            setSupportActionBar(toolbar);
        }

        ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setTitle(R.string.notification_settings);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
        }
    }

    /**
     * Sets up master switch to disable/enable all notification settings
     */
    private void setUpMasterSwitch() {
        PrefMasterSwitchToolbarView masterSwitchToolBarView = findViewById(R.id.master_switch);
        masterSwitchToolBarView.setMasterSwitchToolbarListener(this);

        // Set master switch state from shared preferences.
        boolean isMasterChecked = mSharedPreferences.getBoolean(getString(R.string.wp_pref_notifications_master), true);
        masterSwitchToolBarView.loadInitialState(isMasterChecked);

        ElevationOverlayProvider elevationOverlayProvider = new ElevationOverlayProvider(this);
        float cardElevation = getResources().getDimension(R.dimen.card_elevation);
        int appBarColor = elevationOverlayProvider
                .compositeOverlay(ContextExtensionsKt.getColorFromAttribute(this, R.attr.wpColorAppBar), cardElevation);

        masterSwitchToolBarView.setBackgroundColor(appBarColor);

        hideDisabledView(isMasterChecked);
    }

    @Override
    public void onMasterSwitchCheckedChanged(
            CompoundButton buttonView,
            boolean isChecked
    ) {
        mSharedPreferences.edit().putBoolean(getString(R.string.wp_pref_notifications_master), isChecked)
                          .apply();

        hideDisabledView(isChecked);

        if (isChecked) {
            AnalyticsTracker.track(AnalyticsTracker.Stat.NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_ENABLED);
        } else {
            AnalyticsTracker.track(AnalyticsTracker.Stat.NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_DISABLED);
        }
    }

    /**
     * Hide view when Notification Settings are disabled by toggling the master switch off.
     *
     * @param isMasterChecked TRUE to hide disabled view, FALSE to show disabled view
     */
    protected void hideDisabledView(boolean isMasterChecked) {
        LinearLayout notificationsDisabledView = findViewById(R.id.notification_settings_disabled_view);
        notificationsDisabledView.setVisibility(isMasterChecked ? View.INVISIBLE : View.VISIBLE);
        mFragmentContainer.setVisibility(isMasterChecked ? View.VISIBLE : View.GONE);
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/notifications/NotificationsSettingsActivity.kt
package org.wordpress.android.ui.prefs.notifications

import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.widget.CompoundButton
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.preference.PreferenceManager
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode.MAIN
import org.wordpress.android.R.id
import org.wordpress.android.R.layout
import org.wordpress.android.R.string
import org.wordpress.android.analytics.AnalyticsTracker
import org.wordpress.android.analytics.AnalyticsTracker.Stat.NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_DISABLED
import org.wordpress.android.analytics.AnalyticsTracker.Stat.NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_ENABLED
import org.wordpress.android.modules.APPLICATION_SCOPE
import org.wordpress.android.ui.LocaleAwareActivity
import org.wordpress.android.ui.notifications.NotificationEvents.NotificationsSettingsStatusChanged
import org.wordpress.android.ui.prefs.notifications.PrefMainSwitchToolbarView.MainSwitchToolbarListener
import org.wordpress.android.ui.prefs.notifications.usecase.UpdateNotificationSettingsUseCase
import javax.inject.Inject
import javax.inject.Named

@AndroidEntryPoint
class NotificationsSettingsActivity : LocaleAwareActivity(), MainSwitchToolbarListener {
    @Inject
    lateinit var updateNotificationSettingsUseCase: UpdateNotificationSettingsUseCase

    @Inject
    @Named(APPLICATION_SCOPE)
    lateinit var applicationScope: CoroutineScope

    private lateinit var messageTextView: TextView
    private lateinit var messageContainer: View

    private lateinit var fragmentContainer: View

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout.notifications_settings_activity)
        fragmentContainer = findViewById(id.fragment_container)

        setUpPrimaryToolbar()

        setUpMainSwitch()

        if (savedInstanceState == null) {
            @Suppress("DEPRECATION")
            fragmentManager.beginTransaction()
                .add(id.fragment_container, NotificationsSettingsFragment())
                .commit()
        }

        messageContainer = findViewById(id.notifications_settings_message_container)
        messageTextView = findViewById(id.notifications_settings_message)
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        super.onStop()
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressedDispatcher.onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    @Subscribe(threadMode = MAIN)
    fun onEventMainThread(event: NotificationsSettingsStatusChanged) {
        if (TextUtils.isEmpty(event.message)) {
            messageContainer.visibility = View.GONE
        } else {
            messageContainer.visibility = View.VISIBLE
            messageTextView.text = event.message
        }
    }

    /**
     * Set up primary toolbar for navigation and search
     */
    private fun setUpPrimaryToolbar() {
        val toolbar = findViewById<Toolbar>(id.toolbar_with_search)

        toolbar?.let { setSupportActionBar(it) }

        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowTitleEnabled(false)
        }
    }

    /**
     * Sets up main switch to disable/enable all notification settings
     */
    private fun setUpMainSwitch() {
        val mainSwitchToolBarView = findViewById<PrefMainSwitchToolbarView>(id.main_switch)
        mainSwitchToolBarView.setMainSwitchToolbarListener(this)

        // Set main switch state from shared preferences.
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this@NotificationsSettingsActivity)
        val isMainChecked = sharedPreferences.getBoolean(getString(string.wp_pref_notifications_main), true)
        mainSwitchToolBarView.loadInitialState(isMainChecked)
        hideDisabledView(isMainChecked)
    }

    override fun onMainSwitchCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
        applicationScope.launch { updateNotificationSettingsUseCase.updateNotificationSettings(isChecked) }

        hideDisabledView(isChecked)

        if (isChecked) {
            AnalyticsTracker.track(NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_ENABLED)
        } else {
            AnalyticsTracker.track(NOTIFICATION_SETTINGS_APP_NOTIFICATIONS_DISABLED)
        }
    }

    /**
     * Hide view when Notification Settings are disabled by toggling the main switch off.
     *
     * @param isMainChecked TRUE to hide disabled view, FALSE to show disabled view
     */
    private fun hideDisabledView(isMainChecked: Boolean) {
        val notificationsDisabledView = findViewById<LinearLayout>(id.notification_settings_disabled_view)
        notificationsDisabledView.visibility = if (isMainChecked) View.INVISIBLE else View.VISIBLE
        fragmentContainer.visibility = if (isMainChecked) View.VISIBLE else View.GONE
    }
}
